Update Dependency Management section of docs - Add advanced tutorial (#32511)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/dep-man/advanced/artifact_resolution.adoc

        "org.gradle.dependency.bundling": "external",
        "org.gradle.jvm.version": "8",
        "org.gradle.libraryelements": "jar",
        "org.gradle.usage": "java-api"
      },
      "files": [
        {
          "name": "kotlin-stdlib-1.8.10.jar"
        }
      ]
    },
    {
      "name": "runtimeElements",
      "attributes": {
        "org.gradle.category": "library",
        "org.gradle.dependency.bundling": "external",
        "org.gradle.jvm.version": "8",
        "org.gradle.libraryelements": "jar",
        "org.gradle.usage": "java-runtime"
      },
      "files": [
        {
          "name": "kotlin-stdlib-1.8.10.jar"
        }
      ]
    },
    {
      "name": "jdk7ApiElements",
      "attributes": {
        "org.gradle.usage": "java-api",
        "org.gradle.jvm.version": "7"
      },
      "files": [
        {
          "name": "kotlin-stdlib-jdk7-1.8.10.jar"
        }
      ]
    },
    {
      "name": "jdk8ApiElements",
      "attributes": {
        "org.gradle.usage": "java-api",
        "org.gradle.jvm.version": "8"
      },
      "files": [
        {
          "name": "kotlin-stdlib-jdk8-1.8.10.jar"
        }
      ]
    }
  ]
}
----

> As we can see there are a number of *artifacts* available in the *metadata*:

Author: Added.

[cols="1,1"]
|===
| Variant | Artifact(s) | Purpose

| `apiElements`
| `kotlin-stdlib-1.8.10.jar`
| Standard Kotlin runtime (default).

| `runtimeElements`
| `kotlin-stdlib-jdk7-1.8.10.jar`
| Provides additional APIs for Java 7 compatibility.

| `jdk7ApiElements`
| `kotlin-stdlib-jdk8-1.8.10.jar`
| Provides additional APIs for Java 8 compatibility.

| `jdk8ApiElements`
| `kotlin-stdlib-common-1.8.10.jar`
| Shared standard library for Kotlin Multiplatform.
|===

> In an ideal case, once a *variant* is selected, the *artifact* should be automatically resolved.

Reviewer: ```suggestion
Typically, once a *variant* is selected, its associated *artifacts* can be automatically resolved.
```

However, there are specific reasons why *artifact* selection still happens _after_ *variant* selection.
For example, what if the metadata for `org.jetbrains.kotlin:kotlin-stdlib:1.8.10` looked like this:

[source,json]
.kotlin-stdlib-1.8.10.module
----
{
  "variants": [
    {
      "name": "apiElements",
      "attributes": {
        "org.gradle.category": "library",
        "org.gradle.dependency.bundling": "external",
        "org.gradle.jvm.version": "8",
        "org.gradle.libraryelements": "jar",
        "org.gradle.usage": "java-api"
      },

>       "files": [

Reviewer: ❗ I don't think this properly represents multiple Artifact Sets on a variant.  This is just a single variant with multiple artifacts.

Author: Fixed.

        {
          "name": "kotlin-stdlib-1.8.10.jar"
        },
        {
          "name": "kotlin-stdlib-1.8.10.tar"
        }
      ]
    }
  ]
}

----

How would Gradle know which file to download?

== Artifact Sets

An *artifact set* is a group of *artifacts* that belong to a single *variant* of a *module*.

A single *variant* can include multiple *artifact sets*, each serving a different purpose.

Let's look at the `org.jetbrains.kotlin:kotlin-stdlib:1.8.10` example:

[cols="2,1,~"]
|===
| Variant | Set | Artifacts in the Set

| `apiElements`
| 1
| jar → The packaged library JAR (`kotlin-stdlib-1.8.10.jar`)

|
| 2
| tar → The packaged library TAR (`kotlin-stdlib-1.8.10.tar`)

| `runtimeElements`
| 1
| jar → The packaged library JAR (`kotlin-stdlib-1.8.10.jar`)

|
| 2
| classes → The compiled .class files (`build/classes/java/main/`)

|
| 3
| resources → The associated resource files (`build/resources/main/`)

| `jdk8ApiElements`
| 1
| jar → The packaged library JAR (`kotlin-stdlib-jdk8-1.8.10.jar`)

| `jdk7ApiElements`
| 1
| jar → The packaged library JAR (`kotlin-stdlib-jdk7-1.8.10.jar`)
|===

> The `apiElements` *variant* of `org.jetbrains.kotlin:kotlin-stdlib:1.8.10` provides two *artifact sets*—`jar`, and `tar`—each representing the same distributable in a different form.

Reviewer: 💅 We probably shouldn't use Kotlin as an example here.  It has changed quite a bit, the stdlib shouldn't be directly depended upon, and KMP complicates things.  We could use the java plugins jar vs. classes.  Or make up an artificial example.

Author: I will switch to a different example at a later time.

To perform *artifact* selection (and implicitly, graph resolution) using these default *attributes*, use the `FileCollection` and `ArtifactCollection` APIs.

NOTE: Files can also be accessed from the configuration's `ResolvedConfiguration`, `LenientConfiguration`, `ResolvedArtifact` and `ResolvedDependency` APIs.
However, these APIs are in maintenance mode and are discouraged for use in new development.
These APIs perform artifact selection using the default attributes.

==== Resolving files

To resolve files, we first define a task that accepts a `ConfigurableFileCollection` as input:

====
include::sample[dir="snippets/dependencyManagement/performingResolution-artifactResolution/kotlin",files="build.gradle.kts[tags=file-resolution-task]"]
include::sample[dir="snippets/dependencyManagement/performingResolution-artifactResolution/groovy",files="build.gradle[tags=file-resolution-task]"]
====

Then, we can wire up a resolvable configuration's files to the task's input.
The `Configuration` directly implements `FileCollection` and can be wired directly.

> Alternatively, wiring through `Configuration.getIncoming()` is a more explicit approach:

Reviewer: 💭 I think we should only provide one example here.  Multiple ways to do things is such a large complaint about Gradle.

Author: Fixed.

This process is slightly more complicated, as in order to maintain Configuration Cache compatibility, we need to split the fields of `ResolvedArtifactResult` into two task inputs:

====
include::sample[dir="snippets/dependencyManagement/performingResolution-artifactResolution/kotlin",files="build.gradle.kts[tags=artifact-resolution-task]"]
include::sample[dir="snippets/dependencyManagement/performingResolution-artifactResolution/groovy",files="build.gradle[tags=artifact-resolution-task]"]
====

This task is initialized similarly to the file resolution task:

====
include::sample[dir="snippets/dependencyManagement/performingResolution-artifactResolution/kotlin",files="build.gradle.kts[tags=implicit-artifact-resolution]"]
include::sample[dir="snippets/dependencyManagement/performingResolution-artifactResolution/groovy",files="build.gradle[tags=implicit-artifact-resolution]"]
====

Running this task, we can see that file metadata is included in the output:

----
include::{snippetsPath}/dependencyManagement/performingResolution-artifactResolution/tests/resolve-artifacts.out[]
----

[[sec:customizing-artifact-selection]]
=== Custom Artifact Selection

> An link:{javadocPath}/org/gradle/api/artifacts/ArtifactView.html[`ArtifactView`] API operates on top of the resolved graph, defined by the link:{javadocPath}/org/gradle/api/artifacts/result/ResolutionResult.html[`ResolutionResult`].

Reviewer: ```suggestion
The link:{javadocPath}/org/gradle/api/artifacts/ArtifactView.html[`ArtifactView`] API operates on top of the resolved graph, defined by a {javadocPath}/org/gradle/api/artifacts/result/ResolutionResult.html[`ResolutionResult`].
```
